src/Display/Modal/Modal.tsx

```diff
@@ -29,22 +37,61 @@ const Modal = (props: Props) => {
     ...restProps
   } = props;
 
+  const modalContainerRef = React.useRef(null);
   const modalContentAreaRef = React.useRef(null);
 
+  // mountingState controls mounting and animation
+  const [mountingState, setMountingState] = React.useState(() => {
+    return isVisible ? MountingState.VISIBLE : MountingState.UNMOUNTED;
+  });
+
+  // isOpen set css transition styles
+  const isOpen = mountingState === MountingState.VISIBLE;
+
+  // MOUNTED mount modal, but modal is not visible
+  React.useEffect(() => {
+    if (isVisible && mountingState === MountingState.UNMOUNTED) {
+      setMountingState(MountingState.MOUNTED);
+    }
+  }, [isVisible, mountingState]);
+
+  // VISIBLE set isOpen to true, triggers css transition
+  React.useEffect(() => {
+    if (isVisible && mountingState === MountingState.MOUNTED) {
+      setMountingState(MountingState.VISIBLE);
+    }
+  }, [isVisible, mountingState]);
+
+  // UNMOUNT_START set isOpen to false, triggers css transition
   React.useEffect(() => {
-    if (isVisible) {
+    if (!isVisible && mountingState === MountingState.VISIBLE) {
+      setMountingState(MountingState.UNMOUNT_START);
+    }
+  }, [isVisible, mountingState]);
```

Comment:
Should we combine those 3 `useEffect` above? Since they have the same dependency.

Reply:
Do you think make them as one effect will be easier to read?

Reply:
ok